flow/flow.go

```diff
@@ -0,0 +1,412 @@
+	ConsentSkip        bool             `json:"consent_skip" db:"consent_skip"`
+	ConsentVerifier    string           `json:"-" db:"consent_verifier"`
+	ConsentCSRF        string           `json:"-" db:"consent_csrf"`
+
+	CHGrantedScope       sqlxx.StringSlicePipeDelimiter `json:"ch_grant_scope" db:"ch_granted_scope"`
+	CHGrantedAudience    sqlxx.StringSlicePipeDelimiter `json:"ch_grant_access_token_audience" db:"ch_granted_at_audience"`
+	CHRemember           bool                           `json:"ch_remember" db:"ch_remember"`
+	CHRememberFor        int                            `json:"ch_remember_for" db:"ch_remember_for"`
+	CHHandledAt          sqlxx.NullTime                 `json:"ch_handled_at" db:"ch_handled_at"`
+	CHWasHandled         bool                           `json:"-" db:"ch_was_used"`
+	CHError              *consent.RequestDeniedError    `json:"-" db:"ch_error"`
+	CHSessionIDToken     sqlxx.MapStringInterface       `json:"-" db:"ch_session_id_token" faker:"-"`
+	CHSessionAccessToken sqlxx.MapStringInterface       `json:"-" db:"ch_session_access_token" faker:"-"`
+}
+
+func NewFlow(r *consent.LoginRequest) *Flow {
+	return &Flow{
+		ID:                     r.ID,
+		RequestedScope:         r.RequestedScope,
+		RequestedAudience:      r.RequestedAudience,
+		Skip:                   r.Skip,
+		Subject:                r.Subject,
+		OpenIDConnectContext:   r.OpenIDConnectContext,
+		Client:                 r.Client,
+		ClientID:               r.ClientID,
+		RequestURL:             r.RequestURL,
+		SessionID:              r.SessionID,
+		WasHandled:             r.WasHandled,
+		ForceSubjectIdentifier: r.ForceSubjectIdentifier,
+		Verifier:               r.Verifier,
+		CSRF:                   r.CSRF,
+		LoginAuthenticatedAt:   r.AuthenticatedAt,
+		RequestedAt:            r.RequestedAt,
+		State:                  FlowStateLoginInitialized,
+	}
+}
+
+func (f *Flow) HandleLoginRequest(h *consent.HandledLoginRequest) error {
+	// TODO Validate HandledLoginRequest. This will require updating tests that expect the flow to fail at a later point.
+	if f.State != FlowStateLoginInitialized {
+		return errors.Errorf("invalid flow state: expected %d, got %d", FlowStateLoginInitialized, f.State)
+	}
+	if f.ID != h.ID {
+		return errors.Errorf("flow ID %s does not match HandledLoginRequest ID %s", f.ID, h.ID)
+	}
+	if f.Subject != "" && h.Subject != "" && f.Subject != h.Subject {
+		return errors.Errorf("flow Subject %s does not match the HandledLoginRequest Subject %s", f.Subject, h.Subject)
+	}
+	if f.ForceSubjectIdentifier != "" && h.ForceSubjectIdentifier != "" && f.ForceSubjectIdentifier != h.ForceSubjectIdentifier {
+		return errors.Errorf("flow ForceSubjectIdentifier %s does not match the HandledLoginRequest ForceSubjectIdentifier %s", f.ForceSubjectIdentifier, h.ForceSubjectIdentifier)
+	}
+
+	f.State = FlowStateLoginUnused // TODO FlowStateError if h.Error != nil
```

Comment:
TODO?

Reply:
Copying my comment from the other thread: "In order to benefit from the error states (rather than adding yet another thing that needs to be checked when working with errors), we'd need to change a lot of the error handling logic. But this is PR is about refactoring the database, so I decided to leave it as a TODO. Should I remove the error states completely and just add a note that they could be introduced in the future?"

Reply:
Yes, add a short note :)